### PR TITLE
feat(lsp): Add textDocument/implementation

### PR DESCRIPTION
**Pull request by matsujika**
Part of #8643

### CODE REVIEW

File: cli/lsp/tsc.rs [hunk 1]
```diff
@@ -541,34 +598,18 @@ impl DefinitionInfoAndBoundSpan {
...
-              )
-            };
-          location_links.push(lsp_types::LocationLink {
-            origin_selection_range: Some(self.text_span.to_range(line_index)),
-            target_uri,
-            target_range,
-            target_selection_range,
-          });
+        if let Some(link) = di
+          .document_span
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:608]: Nice! 💯

**Review by kitsonk — APPROVED** [review]
Nice contribution... Thank you.

LGTM

We need to enable "ignoring" the same results from the built in language service in the vscode plugin: https://github.com/denoland/vscode_deno/issues/301.  I can do that if you wish, but also feel free to raise a PR for that.

**Comment by matsujika** [comment]
> We need to enable "ignoring" the same results from the built in language service in the vscode plugin: denoland/vscode_deno#301. I can do that if you wish, but also feel free to raise a PR for that.

Can you do that instead of me please? I unfortunately have exams this and next week.

**Comment by rwols** [comment]
@matsujika when opening a remote file with textDocument/implementation, the server now returns an URI with a strange scheme:

```
:: --> Deno textDocument/implementation(4): {'position': {'character': 12, 'line': 0}, 'textDocument': {'uri': 'file:///Users/raoulwols/dev/denoserver/server.ts'}, 'workDoneToken': 'wd4'}
:: <<< Deno 4: [{'targetUri': 'deno:/https/deno.land/std/http/server.ts', 'targetRange': {'end': {'character': 1, 'line': 305}, 'start': {'character': 0, 'line': 298}}, 'targetSelectionRange': {'end': {'character': 21, 'line': 298}, 'start': {'character': 16, 'line': 298}}}]
```

I would expect `targetUri` to start with `https://`. Instead it starts with `deno:/`. What is the reasoning behind using `deno:/`?

**Comment by kitsonk** [comment]
`deno:` is used so that IDEs can ensure that the Deno LSP handles the documents as virtual text documents. It is necessary for the proper operation of LSP.

**Comment by rwols** [comment]
This seems to require client glue code for each separate client, which defeats the purpose of LSP. https://github.com/denoland/vscode_deno/blob/34b7130f3b810a635f88dd88e5927f3faaa1bffa/client/src/lsp_extensions.ts

Handling custom URI schemes is something we can account for I guess.

**Comment by kitsonk** [comment]
@rwols far from it. There is always still some client glue no matter what, for commands. Just passing the raw scheme would cause lots of other client issues, because for remote modules. Like you don't require extensions on URLs so how does the editor know the media type to route to the correct extension. By moving it to a custom scheme it is 100% clear to the editor who additional requests and notifications should go through. You could maybe argue it is a gap in the LSP protocol, but the protocol is designed to have extensions.

Those lsp extensions are documented in the README for the Deno LSP: https://github.com/denoland/deno/tree/main/cli/lsp

**Comment by rwols** [comment]
The choice could have been made to apply the following strategy:

1. Download the resource to a temp location on the filesystem.
2. Set `targetUri` to that temp location on the filesystem.
3. Reply to the client request with a `file://` scheme.

Not every client has the notion of a "virtual text document".

**Comment by kitsonk** [comment]
@rowls that does not work... we previously tried that in the v2 branch of the extension. URLs don't make great file systems paths, especially when accounting for queries, also consider data URLs which Deno supports.  That means you have to hash the filename and then the hashed file name doesn't look anything like the source file, so you get these really ugly documents.  The virtual document strategy is far more robust.  Would be glad to work with any team trying to build a client for Deno LSP to understand how virtual text document support is critical, as remote modules are virtual documents that can be opened in the editor.

**Comment by lucacasonato** [comment]
@rwols Could you make a new issue for that? It is hard to keep track of things if they are just posted as comments on old PRs :-)

**Pull request merged**